src/platforms/android/server/graphic_buffer_allocator.cpp

```diff
@@ -138,3 +146,254 @@ std::shared_ptr<mg::Buffer> mga::GraphicBufferAllocator::alloc_buffer(
+
+            if (!(mir_buffer = shim->associated_buffer.lock()))
+            {
+                /*
+                 * We've seen this wl_buffer before, but all the WaylandBuffers associated with it
+                 * have been destroyed.
+                 *
+                 * Recreate a new WaylandBuffer to track the new compositor lifetime.
+                 */
+                mir_buffer = std::shared_ptr<WaylandBuffer>{
+                    new WaylandBuffer{
+                        dpy,
+                        buffer,
+                        extensions,
+                        std::move(on_consumed)}};
+                shim->associated_buffer = mir_buffer;
+            }
+        }
+        else
+        {
+            mir_buffer = std::shared_ptr<WaylandBuffer>{
+                new WaylandBuffer{
+                    dpy,
+                    buffer,
+                    extensions,
+                    std::move(on_consumed)}};
+            shim = new DestructionShim;
+            shim->destruction_listener.notify = &on_buffer_destroyed;
+            shim->associated_buffer = mir_buffer;
+
+            wl_resource_add_destroy_listener(buffer, &shim->destruction_listener);
+        }
+
+        mir_buffer->buffer_mutex = shim->mutex;
+        return mir_buffer;
+    }
+
+    ~WaylandBuffer()
+    {
+        if (egl_image != EGL_NO_IMAGE_KHR)
+            extensions->eglDestroyImageKHR(dpy, egl_image);
+
+        std::lock_guard<std::mutex> lock{*buffer_mutex};
+        if (buffer)
+        {
+            wl_resource_queue_event(buffer, WL_BUFFER_RELEASE);
+        }
+    }
+
+    void gl_bind_to_texture() override
+    {
+        std::unique_lock<std::mutex> lock{*buffer_mutex};
+        if (buffer == nullptr)
+        {
+            mir::log_warning("WaylandBuffer::gl_bind_to_texture() called on a destroyed wl_buffer", this);
+            return;
+        }
+        if (egl_image == EGL_NO_IMAGE_KHR)
+        {
+            eglBindAPI(MIR_SERVER_EGL_OPENGL_API);
+
+            const EGLint image_attrs[] =
+                {
+                    EGL_IMAGE_PRESERVED_KHR, EGL_TRUE,
+                    EGL_NONE
+                };
+
+            egl_image = extensions->eglCreateImageKHR(
+                dpy,
+                EGL_NO_CONTEXT,
+                EGL_WAYLAND_BUFFER_WL,
+                buffer,
+                image_attrs);
+
+            if (egl_image == EGL_NO_IMAGE_KHR)
+                BOOST_THROW_EXCEPTION(mg::egl_error("Failed to create EGLImage"));
+
+            on_consumed();
+        }
+        lock.unlock();
+
+        extensions->glEGLImageTargetTexture2DOES(GL_TEXTURE_2D, egl_image);
+    }
+
+    void bind() override
+    {
+        gl_bind_to_texture();
+    }
+
+    void secure_for_render() override
+    {
+    }
+
+    std::shared_ptr<mir::graphics::NativeBuffer> native_buffer_handle() const override
+    {
+        return nullptr;
```

Comment:
this one needs to impemented - right?

Reply:
Its not implemented upstream so i would guess not yet. mesa platform returns nullptr too

Reply:
Humm, i guess you are right, we might need this